pkg/bindinfo/binding_cache.go

```diff
@@ -48,6 +48,101 @@ var GetBindingReturnNilAlways = stringutil.StringerStr("getBindingReturnNilAlway
 // LoadBindingNothing is only for test
 var LoadBindingNothing = stringutil.StringerStr("LoadBindingNothing")
 
+// digestBiMap represents a bidirectional map between noDBDigest and sqlDigest, used to support cross-db binding.
+// One noDBDigest can map to multiple sqlDigests, but one sqlDigest can only map to one noDBDigest.
+type digestBiMap interface {
+	// Add adds a pair of noDBDigest and sqlDigest.
+	// noDBDigest is the digest calculated after eliminating all DB names, e.g. `select * from test.t` -> `select * from t` -> noDBDigest.
+	// sqlDigest is the digest where all DB names are kept, e.g. `select * from test.t` -> exactDigest.
+	Add(noDBDigest, sqlDigest string)
+
+	// Del deletes the pair of noDBDigest and sqlDigest.
+	Del(sqlDigest string)
+
+	// NoDBDigest2SQLDigest converts noDBDigest to sqlDigest.
+	NoDBDigest2SQLDigest(noDBDigest string) []string
+
+	// SQLDigest2NoDBDigest converts sqlDigest to noDBDigest.
+	SQLDigest2NoDBDigest(sqlDigest string) string
+
+	// Copy copies this digestBiMap.
+	Copy() digestBiMap
+}
+
+type digestBiMapImpl struct {
+	mu                   sync.RWMutex
+	noDBDigest2SQLDigest map[string][]string // noDBDigest --> sqlDigests
+	sqlDigest2noDBDigest map[string]string   // sqlDigest --> noDBDigest
+}
+
+func newDigestBiMap() digestBiMap {
+	return &digestBiMapImpl{
+		noDBDigest2SQLDigest: make(map[string][]string),
+		sqlDigest2noDBDigest: make(map[string]string),
+	}
+}
+
+func (b *digestBiMapImpl) Add(noDBDigest, sqlDigest string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.noDBDigest2SQLDigest[noDBDigest] = append(b.noDBDigest2SQLDigest[noDBDigest], sqlDigest)
+	b.sqlDigest2noDBDigest[sqlDigest] = noDBDigest
+}
+
+func (b *digestBiMapImpl) Del(sqlDigest string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	noDBDigest, ok := b.sqlDigest2noDBDigest[sqlDigest]
+	if !ok {
+		return
+	}
+	digestList := b.noDBDigest2SQLDigest[noDBDigest]
+	for i := range digestList { // remove sqlDigest from this list
+		if digestList[i] == sqlDigest {
+			// Deleting binding is a low-frequently operation, so the O(n) performance is enough.
+			digestList = append(digestList[:i], digestList[i+1:]...)
+			break
+		}
+	}
+	if len(digestList) == 0 {
+		delete(b.noDBDigest2SQLDigest, noDBDigest)
+	} else {
+		b.noDBDigest2SQLDigest[noDBDigest] = digestList
+	}
+	delete(b.sqlDigest2noDBDigest, sqlDigest)
+}
+
+func (b *digestBiMapImpl) NoDBDigest2SQLDigest(noDBDigest string) []string {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.noDBDigest2SQLDigest[noDBDigest]
+}
+
+func (b *digestBiMapImpl) SQLDigest2NoDBDigest(sqlDigest string) string {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.sqlDigest2noDBDigest[sqlDigest]
+}
+
+func (b *digestBiMapImpl) Copy() digestBiMap {
```

Comment:
what the case of cp a map here？

Reply:
When we load new bindings, we'll copy the old binding cache and then insert new records into it and then replace it with the old one.